src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -315,6 +324,9 @@ public synchronized Object get(long timeout, TimeUnit unit) throws InterruptedEx
     }
 
     private void setOutcome(Outcome o) {
+        if (bodyToUnexport != null && thread != null) {
```

Comment:
Seems the easiest place to do this—called from both the success and failure adapters.